src/plugins/intel_cpu/src/nodes/subgraph.cpp

```diff
@@ -506,50 +674,114 @@ void Snippet::prepareParams() {
             size_t ndims_to_skip = norm_shape.size() - original_input_shape_ranks[i];
             new_shapes.emplace_back(norm_shape.begin() + ndims_to_skip, norm_shape.end());
         }
-        snippet->reshape_body(new_shapes);
+        snippet_for_generation->reshape_body(new_shapes);
     }
+    snippet_for_generation->set_master_shape(ov::PartialShape(masterShape));
+    snippet_for_generation->set_tile_rank(tileRank);
 
-    snippet->set_master_shape(ov::PartialShape(masterShape));
-    snippet->set_tile_rank(tileRank);
+    // generate
+    jit_snippets_compile_args jcp;
+    jcp.master_shape = masterShape;
+    jcp.tile_rank = tileRank;
+    generate(&jcp);
+    buffer_scratchpad_size = snippet_for_generation->get_buffer_scratchpad_size();
+    buffer_scratchpad.resize(buffer_scratchpad_size * parallel_get_max_threads(), 0);
 }
 
-bool Snippet::needPrepareParams() const {
-    return inputShapesModified() || !schedule.ptr;
+ov::PartialShape Snippet::SnippetJitExecutor::canonicalizeBody() {
+    ov::snippets::op::Subgraph::BlockedShapeVector input_blocked_shapes = getBlockedShapes(
+        snippetAttrs.inMemBlockedDims, snippetAttrs.inMemOrders, snippetAttrs.inMemPrecs);
+
+    ov::snippets::op::Subgraph::BlockedShapeVector output_blocked_shapes = getBlockedShapes(
+        snippetAttrs.outMemBlockedDims, snippetAttrs.outMemOrders, snippetAttrs.outMemPrecs);
+
+    const auto& canonicalShape = snippetAttrs.snippet->canonicalize(output_blocked_shapes, input_blocked_shapes);
+    return canonicalShape;
 }
 
-bool Snippet::canBeInPlace() const {
-    if (isDynamic || getParentEdgesAtPort(0)[0]->getParent()->getType() == Type::Input) {
-        return false;
-    }
-    if (getChildEdges().size() != 1) {
+ov::PartialShape Snippet::SnippetJitExecutor::reshapeCanonicalizedBody() {
+    ov::snippets::op::Subgraph::BlockedShapeVector input_blocked_shapes = getBlockedShapes(
+        snippetAttrs.inMemBlockedDims, snippetAttrs.inMemOrders, snippetAttrs.inMemPrecs);
+
+    ov::snippets::op::Subgraph::BlockedShapeVector output_blocked_shapes = getBlockedShapes(
+        snippetAttrs.outMemBlockedDims, snippetAttrs.outMemOrders, snippetAttrs.outMemPrecs);
+
+    const auto& canonicalShape = snippetAttrs.snippet->reshape_canonicalized_body(output_blocked_shapes, input_blocked_shapes);
```

Comment:
Seems like we can unite 2 methods `reshapeCanonicalizedBody` and `canonicalizeBody` into one `canonicalizeBody` but just add the parameter `reshape`. If `true`, use `reshape_canonicalized_body`. Otherwise use `canonicalize`.